src/librustc/ty/util.rs

```diff
@@ -525,6 +393,142 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
     }
 }
 
+struct TypeIdHasher<'a, 'gcx: 'a+'tcx, 'tcx: 'a> {
+    tcx: TyCtxt<'a, 'gcx, 'tcx>,
+    svh: &'a Svh,
+    state: SipHasher
+}
+
+impl<'a, 'gcx, 'tcx> TypeIdHasher<'a, 'gcx, 'tcx> {
+    fn hash<T: Hash>(&mut self, x: T) {
+        x.hash(&mut self.state);
+    }
+
+    fn hash_discriminant_u8<T>(&mut self, x: &T) {
+        let v = unsafe {
+            intrinsics::discriminant_value(x)
+        };
+        let b = v as u8;
+        assert_eq!(v, b as u64);
+        self.hash(b)
+    }
+
+    fn def_id(&mut self, did: DefId) {
+        let h = if did.is_local() {
+            self.svh.clone()
+        } else {
+            self.tcx.sess.cstore.crate_hash(did.krate)
+        };
+        self.hash(h);
+        self.hash(did.index);
```

Comment:
That and not hashing the SVH would mean that `TypeId` is stable, which I guess is needed for incremental recompilation, but a bit scary. Could `TypeId` somehow include the `struct`/`enum` definition too? At least the things SVH would include and/or incremental recompilation would consider relevant to all the type's users.

OTOH, what's the worst that can happen? Different versions of the same crate would use the same `TypeId` and someone might be inclined to, e.g. build an IPC bridge between them.
Which would keep believing the types are the same even if the layout has changed.

Stability-wise, do we want to allow that? Seems to me that once we make it possible, people might start depending on it for various crazy hacks.